mtt/src/mtt.rs

```diff
@@ -0,0 +1,688 @@
+// |  63:36: Zero | 35:34: Type | 33:0: Info
+// |---------------------------------------------------
+// The 2-bit encoding for Type is as follows:
+// 00b: Non-confidential 1GB range
+// 01b: Confidential 1GB range
+// 10b: A 64-MB range further described by 16K entries in a 4K page (pointed to by Info[33:0])
+// 11b: A 64-MB range composed of 32x2MB sub-ranges, and further described by Info[31:0]
+enum MttL2Entry<'a> {
+    // The entire 1GB range containing the address is non-confidential.
+    // 1GB ranges are mapped using 64MB subranges, and by convention, the invariant is
+    // that each of the 16 consecutive entries in the L2 table have the same type.
+    NonConfidential1G(Mtt1GEntry<'a, NonConf>),
+    // The entire 1GB range containing the address is confidential.
+    // 1GB ranges are mapped using 64MB subranges, and by convention, the invariant is
+    // that each of the 16 consecutive entries in the L2 table have the same type.
+    Confidential1G(Mtt1GEntry<'a, Conf>),
+    // The 64-MB range has been partitioned into 2MB regions, and each sub-region can be
+    // confidential or non-confidential
+    Mixed2M(Mtt64MEntry<'a>),
+    // The 64-MB range has been partitioned into 16K regions of 4KB size.
+    // Each sub-region can be confidential or non-confidential
+    L1Entry(MttL1Entry<'a>),
+}
+
+struct MttL2Directory {
+    mtt_l2_base: &'static mut [u64],
+}
+
+impl MttL2Directory {
+    fn get_memory_type(addr: SupervisorPageAddr, l2_entry: &MttL2Entry) -> MttMemoryType {
+        use MttL2Entry::*;
+        match l2_entry {
+            Confidential1G(_) => MttMemoryType::Confidential,
+            NonConfidential1G(_) => MttMemoryType::NonConfidential,
+            Mixed2M(mixed_2m_entry) => {
+                if mixed_2m_entry.is_conf(addr) {
+                    MttMemoryType::Confidential
+                } else {
+                    MttMemoryType::NonConfidential
+                }
+            }
+            L1Entry(l1_entry) => {
+                if l1_entry.is_conf(addr) {
+                    MttMemoryType::Confidential
+                } else {
+                    MttMemoryType::NonConfidential
+                }
+            }
+        }
+    }
+
+    fn set_memory_type(addr: SupervisorPageAddr, l2_entry: MttL2Entry, memory_type: MttMemoryType) {
+        use MttL2Entry::*;
+        match l2_entry {
+            Confidential1G(conf_1g) => conf_1g.set_memory_type(memory_type),
+            NonConfidential1G(nc_1g) => nc_1g.set_memory_type(memory_type),
+            Mixed2M(mut mixed_2m_entry) => mixed_2m_entry.set_memory_type(addr, memory_type),
+            L1Entry(mut l1_entry) => l1_entry.set_memory_type(addr, memory_type),
+        }
+    }
+
+    // Returns the index of the MTT L2 entry for the physical address
+    fn get_mtt_index(phys_addr: SupervisorPageAddr) -> usize {
+        let addr = phys_addr.bits();
+        // Bits 45:26
+        ((addr & 0x3fff_fc00_0000) >> 26) as usize
```

Comment:
these magic numbers should be defined in terms of constants, e.g.: `(addr & (MTT_MAX_PA - 1)) >> MTT_L2_SHIFT`